File: gateway/adapters/discord.py
from __future__ import annotations

import logging
from typing import Callable

from .base import ChannelAdapter

log = logging.getLogger(__name__)


class DiscordAdapter(ChannelAdapter):
    """Discord adapter using discord.py."""

    def __init__(self, token: str):
        self._token = token
        self._client = None

    async def start(self, on_message: Callable) -> None:
        import discord

        intents = discord.Intents.default()
        intents.message_content = True
        client = discord.Client(intents=intents)
        self._client = client
        _router_on_message = on_message

        @client.event
        async def on_message(message):
            if message.author == client.user:
                return
            peer_id = str(message.author.id)

            async def deliver(pid: str, text: str) -> None:
                await message.channel.send(text)

            await _router_on_message("discord", peer_id, message.content, deliver)

        log.info("Discord adapter starting")
        await client.start(self._token)

    async def send(self, peer_id: str, text: str) -> None:
        pass  # Discord requires a channel reference; use deliver() in on_message instead

File: src/gateway/adapters/discord.ts
import { ChannelAdapter } from "./base.ts";
import type { OnMessageFn } from "../router.ts";

export class DiscordAdapter extends ChannelAdapter {
  private discordClient: unknown = null;

  constructor(private readonly token: string) {
    super();
  }

  async start(onMessage: OnMessageFn): Promise<void> {
    const { Client, GatewayIntentBits } = await import("discord.js");

    const client = new Client({
      intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildMessages,
        GatewayIntentBits.MessageContent,
        GatewayIntentBits.DirectMessages,
      ],
    });

    this.discordClient = client;

    client.on("messageCreate", async (message: Record<string, unknown>) => {
      const author = message["author"] as Record<string, unknown>;
      if (author["bot"]) return;

      const peerId = String(author["id"]);
      const content = (message["content"] as string | undefined) ?? "";

      const channel = message["channel"] as { send(text: string): Promise<unknown> };

      await onMessage("discord", peerId, content, async (_pid: string, chunk: string) => {
        await channel.send(chunk);
      });
    });

    console.log("Discord adapter starting");
    await client.login(this.token);

    // keep alive
    await new Promise<void>(() => {
      // never resolves — discord.js keeps event loop alive
    });
  }

  async send(_peerId: string, _text: string): Promise<void> {
    // Discord requires a channel reference; use deliver() in onMessage instead
  }
}
